### DISCUSSION TITLE
How can I run for multiple users on localhost?

### DISCUSSION THREAD

**Discussion by RWayne93** [open-webui/open-webui#676] [Q&A]
Do I need to get the docker containers for ollama and webui up and running and then run the kubernetes configuration or is there an easier way to test this out? I just want to see how many clients i can have running on my local machine before it becomes unusuable or too slow and get some metrics for how i would want to set up auto scalling locally. right now just having two users go to localhost the model runs inference synchronously and wont run inference until the first request completes.

**Comment by justinh-rahb** [comment]
Docker remains the easiest way to get started. Syncronous response is expected behaviour with Ollama (and llama.cpp in general), there's no batched inference support. You can work around this somewhat on a large enough Mac by running more instances behind a load-balancer, my 96GB Studio Max can run 2 Ollama instances with Mixtral comfortably, allowing 2 simultaneous users. In practice we rarely have concurrent requests fighting for time, my dozen or so users aren't on it all day, only occasionally. If you're using GPUs, you can run as many Ollamas as you've got GPUs in theory, assigning one to each and load-balancing them for the WebUI backend to connect to.

Future goals for this project include support for more backends beyond llama.cpp, including GPU-specific ones such as vLLM that support batched inference.

**Reply by RWayne93** [reply]
Did llama.cpp not add batch inference when they closed this issue here? 

https://github.com/ggerganov/llama.cpp/issues/3478

**Reply by justinh-rahb** [reply]
From that thread: [ggerganov/llama.cpp/issues/3478#issuecomment-1824047763](https://github.com/ggerganov/llama.cpp/issues/3478#issuecomment-1824047763)

> Ah, I think I got confused. We solved serving clients in parallel, but not processing prompts in parallel.
> A workaround is to submit the prompts in separate requests.

Parallel decoding and batched inference are related but distinct concepts. Parallel decoding generates responses for multiple prompts simultaneously by a single model instance, while batched inference optimizes resource utilization when processing many input sequences as a batch in a single forward pass.

It seems the issue is still being actively debated in Ollama: [ollama/ollama/issues/358](https://github.com/ollama/ollama/issues/358)
